File: db/mongo_init.py

import os
from motor.motor_asyncio import AsyncIOMotorClient
from dotenv import load_dotenv
from logger import logger

# Prefer reading MongoDB-related environment variables from the cloud environment.
# If not available, fall back to reading from the local .env file.
if os.getenv("ENV") == "LOCAL":
    logger.info("Local environment detected. Loading .env file...")
    load_dotenv()  # load local .env file.


MONGO_URI = os.environ.get("MONGO_URI")
DB_NAME = os.environ.get("DB_NAME", "chat_db")
COLLECTION_NAME = os.environ.get("COLLECTION_NAME", "messages")

logger.info(f"MONGO_URI: {MONGO_URI}, DB_NAME: {DB_NAME}, COLLECTION_NAME: {COLLECTION_NAME}")

if not MONGO_URI or not DB_NAME or COLLECTION_NAME:
    logger.error("❌ MONGO_ENV parameters is not set! Please check environment variables.")

# ----------------------------
# MongoDB client
# ----------------------------
mongo_client = AsyncIOMotorClient(MONGO_URI, serverSelectionTimeoutMS=3000)
db = mongo_client[DB_NAME]
messages_collection = db[COLLECTION_NAME]

# ----------------------------
# MongoDB connection check at startup
# ----------------------------
async def check_mongo_connection():
    try:
        # 使用 serverSelectionTimeoutMS 防止卡住
        await mongo_client.admin.command("ping")
        logger.info(f"[✅] MongoDB connection successful. Database '{DB_NAME}' is reachable.")
    except Exception as e:
        logger.error(f"[❌] MongoDB connection failed: {e}")

File: db/mongo_init.go
package db

import (
	"context"
	"log"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

var (
	dBName         string
	collectionName string
	MongoClient    *mongo.Client
)

func InitMongo() error {
	// load mongo env parameters
	mongoURI := os.Getenv("MONGO_URI")
	dBName := os.Getenv("DB_NAME")
	collectionName := os.Getenv("COLLECTION_NAME")

	log.Printf("MONGO_URI: %s, DB_NAME: %s, COLLECTION_NAME: %s", mongoURI, dBName, collectionName)

	if mongoURI == "" || dBName == "" || collectionName == "" {
		log.Fatal("[❌] MONGO_ENV parameters is not set! Please check environment variables.")
	}

	// connect to MongoDB
	mongoClient, err := createMongoConnection(mongoURI)
	if err != nil {
		log.Fatalf("[❌] MongoDB connect error: %v", err)
	}

	log.Printf("[✅] Create a Mongo connection to address: %s, DBname: %s, Collection: %s", mongoURI, dBName, collectionName)

	// check connection
	if err := CheckMongoConnection(mongoClient); err != nil {
		log.Fatalf("[❌] MongoDB connection failed: %v", err)
	}

	// init mongo collection
	if err := InitMongoCollection(mongoClient, dBName, collectionName); err != nil {
		log.Fatalf("[❌] MongoDB collection init failed: %v", err)
	}
	MongoClient = mongoClient
	return nil
}

// create a Mongo Connection
func createMongoConnection(mongoURI string) (*mongo.Client, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	mongoClient, err := mongo.Connect(ctx, options.Client().ApplyURI(mongoURI))
	if err != nil {
		return nil, err
	}
	return mongoClient, nil
}

// Check Mongo Connection
func CheckMongoConnection(mongoClient *mongo.Client) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	return mongoClient.Ping(ctx, nil)
}

// Init Mongo Collection, checks and creates collection if not exists, then rechecks connection
func InitMongoCollection(mongoClient *mongo.Client, dbName string, collectionName string) error {
	db := mongoClient.Database(dbName)
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	collections, err := db.ListCollectionNames(ctx, map[string]interface{}{})
	if err != nil {
		return err
	}
	found := false
	for _, c := range collections {
		if c == collectionName {
			found = true
			break
		}
	}
	if !found {
		if err := db.CreateCollection(ctx, collectionName); err != nil {
			return err
		}
		log.Println("[✅] Mongo messages collection created!")
	} else {
		log.Println("[ℹ️] Mongo messages collection already exists!")
	}
	// recheck connection
	return CheckMongoConnection(mongoClient)
}
